// @nerimity/nerimity.js/build/services/MessageService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchMessage = fetchMessage;
exports.postMessage = postMessage;
exports.editMessage = editMessage;
exports.deleteMessage = deleteMessage;
exports.buttonClickCallback = buttonClickCallback;
exports.request = request;
const serviceEndpoints_1 = require("./serviceEndpoints");
async function fetchMessage(opts) {
    return await request({
        client: opts.client,
        url: serviceEndpoints_1.ServiceEndpoints.EditMessage(opts.channelId, opts.messageId),
        method: "GET",
        useToken: true,
    });
}
async function postMessage(opts) {
    return await request({
        client: opts.client,
        url: serviceEndpoints_1.ServiceEndpoints.PostMessage(opts.channelId),
        method: "POST",
        body: {
            content: opts.content,
            nerimityCdnFileId: opts.nerimityCdnFileId,
            htmlEmbed: opts.htmlEmbed,
            buttons: opts.buttons,
            silent: opts.silent,
            mentionReplies: opts.mentionReplies,
            replyToMessageIds: opts.replyToMessageIds,
        },
        useToken: true,
    });
}
function editMessage(opts) {
    return request({
        client: opts.client,
        url: serviceEndpoints_1.ServiceEndpoints.EditMessage(opts.channelId, opts.messageId),
        method: "PATCH",
        body: {
            content: opts.content,
            htmlEmbed: opts.htmlEmbed,
            buttons: opts.buttons,
        },
        useToken: true,
    });
}
function deleteMessage(opts) {
    return request({
        client: opts.client,
        url: serviceEndpoints_1.ServiceEndpoints.EditMessage(opts.channelId, opts.messageId),
        method: "DELETE",
        useToken: true,
    });
}
function buttonClickCallback(opts) {
    return request({
        client: opts.client,
        url: serviceEndpoints_1.ServiceEndpoints.ButtonClickCallback(opts.channelId, opts.messageId, opts.buttonId),
        method: "POST",
        body: {
            ...(opts.data ? opts.data : {}),
            userId: opts.userId,
        },
        useToken: true,
    });
}
async function request(opts) {
    const stackCapture = {};
    Error.captureStackTrace(stackCapture, request);
    const url = new URL(opts.url);
    url.search = new URLSearchParams(opts.params || {}).toString();
    const response = await fetch(url, {
        method: opts.method,
        body: JSON.stringify(opts.body),
        headers: {
            "Content-Type": "application/json",
            Authorization: opts.useToken ? opts.client.token : "",
        },
    }).catch((err) => {
        throw new Error("Could not connect to server. " + err.message);
    });
    const text = await response.text();
    if (opts.notJSON)
        return text;
    try {
        const json = JSON.parse(text);
        if (!response.ok) {
            throw new Error(JSON.stringify(json));
        }
        return json;
    }
    catch (err) {
        if (err instanceof Error) {
            // Remove the first line of the captured stack ("Error")
            // and append the rest to the real error
            const originalStack = stackCapture.stack?.split("\n").slice(1).join("\n");
            err.stack += "\n    -- caused by --\n" + originalStack;
        }
        throw err;
    }
}
