// @nerimity/nerimity.js/build/services/PostsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPosts = getPosts;
exports.postPost = postPost;
exports.editPost = editPost;
exports.deletePost = deletePost;
const MessageService_1 = require("./MessageService");
const serviceEndpoints_1 = require("./serviceEndpoints");
async function getPosts(client) {
    return await (0, MessageService_1.request)({
        client: client,
        url: serviceEndpoints_1.ServiceEndpoints.GetPosts(),
        method: "GET",
        useToken: true,
    });
}
async function postPost(opts) {
    return await (0, MessageService_1.request)({
        client: opts.client,
        url: serviceEndpoints_1.ServiceEndpoints.PostPost(),
        method: "POST",
        body: {
            content: opts.content,
            nerimityCdnFileId: opts.nerimityCdnFileId,
            poll: opts.poll,
        },
        useToken: true,
    });
}
async function editPost(opts) {
    return await (0, MessageService_1.request)({
        client: opts.client,
        url: serviceEndpoints_1.ServiceEndpoints.EditPost(opts.postId),
        method: "PATCH",
        body: { content: opts.content },
        useToken: true,
    });
}
async function deletePost(opts) {
    return await (0, MessageService_1.request)({
        client: opts.client,
        url: serviceEndpoints_1.ServiceEndpoints.DeletePost(opts.postId),
        method: "DELETE",
        useToken: true,
    });
}
