// @nerimity/nerimity.js/build/services/ServerService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.banServerMember = banServerMember;
exports.unbanServerMember = unbanServerMember;
exports.kickServerMember = kickServerMember;
const MessageService_1 = require("./MessageService");
const serviceEndpoints_1 = require("./serviceEndpoints");
async function banServerMember(client, serverId, userId, reason) {
    return await (0, MessageService_1.request)({
        client: client,
        url: serviceEndpoints_1.ServiceEndpoints.serverMemberBan(serverId, userId),
        method: "POST",
        useToken: true,
        body: { reason },
    });
}
async function unbanServerMember(client, serverId, userId) {
    return await (0, MessageService_1.request)({
        client: client,
        url: serviceEndpoints_1.ServiceEndpoints.serverMemberBan(serverId, userId),
        method: "DELETE",
        useToken: true,
    });
}
async function kickServerMember(client, serverId, userId) {
    return await (0, MessageService_1.request)({
        client: client,
        url: serviceEndpoints_1.ServiceEndpoints.serverMemberKick(serverId, userId),
        method: "DELETE",
        useToken: true,
    });
}
