// @nerimity/nerimity.js/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RolePermissions = exports.ClientUser = exports.Collection = exports.User = exports.Server = exports.ServerRole = exports.ServerMember = exports.Post = exports.MessageButton = exports.Message = exports.ServerChannel = exports.Channel = exports.WebhookBuilder = exports.RPCClient = exports.AttachmentBuilder = exports.Events = exports.Client = void 0;
var Client_1 = require("./classes/Client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return Client_1.Client; } });
Object.defineProperty(exports, "Events", { enumerable: true, get: function () { return Client_1.Events; } });
var Attachment_1 = require("./classes/Attachment");
Object.defineProperty(exports, "AttachmentBuilder", { enumerable: true, get: function () { return Attachment_1.AttachmentBuilder; } });
var RPCClient_1 = require("./classes/RPCClient");
Object.defineProperty(exports, "RPCClient", { enumerable: true, get: function () { return RPCClient_1.RPCClient; } });
var Webhooks_1 = require("./classes/Webhooks");
Object.defineProperty(exports, "WebhookBuilder", { enumerable: true, get: function () { return Webhooks_1.WebhookBuilder; } });
var Channel_1 = require("./classes/Channel");
Object.defineProperty(exports, "Channel", { enumerable: true, get: function () { return Channel_1.Channel; } });
var ServerChannel_1 = require("./classes/ServerChannel");
Object.defineProperty(exports, "ServerChannel", { enumerable: true, get: function () { return ServerChannel_1.ServerChannel; } });
var Message_1 = require("./classes/Message");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return Message_1.Message; } });
var MessageButton_1 = require("./classes/MessageButton");
Object.defineProperty(exports, "MessageButton", { enumerable: true, get: function () { return MessageButton_1.MessageButton; } });
var Post_1 = require("./classes/Post");
Object.defineProperty(exports, "Post", { enumerable: true, get: function () { return Post_1.Post; } });
var ServerMember_1 = require("./classes/ServerMember");
Object.defineProperty(exports, "ServerMember", { enumerable: true, get: function () { return ServerMember_1.ServerMember; } });
var ServerRole_1 = require("./classes/ServerRole");
Object.defineProperty(exports, "ServerRole", { enumerable: true, get: function () { return ServerRole_1.ServerRole; } });
var Server_1 = require("./classes/Server");
Object.defineProperty(exports, "Server", { enumerable: true, get: function () { return Server_1.Server; } });
var User_1 = require("./classes/User");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return User_1.User; } });
var Collection_1 = require("./classes/Collection");
Object.defineProperty(exports, "Collection", { enumerable: true, get: function () { return Collection_1.Collection; } });
var ClientUser_1 = require("./classes/ClientUser");
Object.defineProperty(exports, "ClientUser", { enumerable: true, get: function () { return ClientUser_1.ClientUser; } });
var bitwise_1 = require("./bitwise");
Object.defineProperty(exports, "RolePermissions", { enumerable: true, get: function () { return bitwise_1.RolePermissions; } });
