// @nerimity/nerimity.js/build/classes/Attachment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttachmentBuilder = void 0;
const url = "https://cdn.nerimity.com";
const uploadUrl = `${url}/upload`;
const saveUrl = `${url}/attachments`;
class AttachmentBuilder {
    constructor(file, name) {
        this.file = file;
        this.name = name;
    }
    async build(channel) {
        const formData = new FormData();
        formData.set("file", this.file, this.name);
        const response = await upload(formData);
        await sendUploadChannel(channel.id, response);
        return response.fileId;
    }
}
exports.AttachmentBuilder = AttachmentBuilder;
async function sendUploadChannel(id, cdn) {
    const response = await fetch(`${saveUrl}/${id}/${cdn.fileId}`, {
        method: "POST",
        body: JSON.stringify(cdn),
    });
    if (!response.ok) {
        throw new Error(`Failed to send attachment: ${response.statusText}`);
    }
    return await response.json();
}
async function upload(dat) {
    const response = await fetch(uploadUrl, {
        method: "POST",
        body: dat,
    });
    if (!response.ok) {
        throw new Error(`Failed to upload attachment: ${response.statusText}`);
    }
    return await response.json();
}
