// @nerimity/nerimity.js/build/classes/Channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Channel = void 0;
const AsyncFunctionQueue_1 = require("../AsyncFunctionQueue");
const MessageService_1 = require("../services/MessageService");
const Message_1 = require("./Message");
class Channel {
    constructor(client, channel) {
        this.messageSendQueue = new AsyncFunctionQueue_1.AsyncFunctionQueue();
        this.client = client;
        this.id = channel.id;
        this.type = channel.type;
        this.createdAt = channel.createdAt;
        this.lastMessagedAt = channel.lastMessagedAt;
        if (channel.serverId)
            this.server = this.client.servers.cache.get(channel.serverId);
    }
    async send(content, opts) {
        return this.messageSendQueue.add(async () => {
            const RawMessage = await (0, MessageService_1.postMessage)({
                client: this.client,
                channelId: this.id,
                content: content,
                silent: opts?.silent,
                nerimityCdnFileId: opts?.nerimityCdnFileId,
                htmlEmbed: opts?.htmlEmbed,
                buttons: opts?.buttons,
                replyToMessageIds: opts?.replyToMessageIds,
                mentionReplies: opts?.mentionReplies,
            });
            const message = new Message_1.Message(this.client, RawMessage);
            return message;
        });
    }
    toString() {
        return `[#:${this.id}]`;
    }
    async deleteMessage(messageId) {
        return (0, MessageService_1.deleteMessage)({
            channelId: this.id,
            client: this.client,
            messageId: messageId,
        });
    }
}
exports.Channel = Channel;
