// @nerimity/nerimity.js/build/classes/Client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.Events = void 0;
const eventemitter3_1 = __importDefault(require("eventemitter3"));
const socket_io_client_1 = require("socket.io-client");
const EventNames_1 = require("../EventNames");
const serviceEndpoints_1 = require("../services/serviceEndpoints");
const ApplicationService_1 = require("../services/ApplicationService");
const MessageButton_1 = require("./MessageButton");
const ClientUser_1 = require("./ClientUser");
const User_1 = require("./User");
const Server_1 = require("./Server");
const Post_1 = require("./Post");
const ServerChannel_1 = require("./ServerChannel");
const Message_1 = require("./Message");
const Channels_1 = require("./Channels");
const Reaction_1 = require("./Reaction");
exports.Events = EventNames_1.ClientEvents;
class Client extends eventemitter3_1.default {
    constructor(opts) {
        super();
        if (opts?.apiUrlOverride) {
            (0, serviceEndpoints_1.updatePath)(opts.apiUrlOverride);
        }
        if (opts?.wsUrlOverride) {
            (0, serviceEndpoints_1.updateWsPath)(opts.wsUrlOverride);
        }
        this.socket = (0, socket_io_client_1.io)(serviceEndpoints_1.wsPath, {
            transports: ["websocket"],
            autoConnect: false,
        });
        this.channels = new Channels_1.Channels(this);
        this.users = new User_1.Users(this);
        this.servers = new Server_1.Servers(this);
        this.posts = new Post_1.Posts(this);
        this.messages = new Message_1.Messages(this, opts?.messageCacheLimit);
        new EventHandlers(this);
    }
    updateCommands(token, commands) {
        return (0, ApplicationService_1.updateCommands)(token, commands);
    }
    login(token) {
        this.token = token;
        this.socket.connect();
    }
}
exports.Client = Client;
class EventHandlers {
    constructor(client) {
        this.client = client;
        this.socket = client.socket;
        client.socket.on(EventNames_1.SocketServerEvents.CONNECT, this.onConnect.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.USER_AUTHENTICATED, this.onAuthenticated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.FRIEND_REQUEST_ACCEPTED, this.onAuthenticated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_MEMBER_JOINED, this.onServerMemberJoined.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_MEMBER_UPDATED, this.onServerMemberUpdated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_MEMBER_LEFT, this.onServerMemberLeft.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_JOINED, this.onServerJoined.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_CHANNEL_CREATED, this.onServerChannelCreated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_CHANNEL_UPDATED, this.onServerChannelUpdated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_CHANNEL_DELETED, this.onServerChannelDeleted.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_LEFT, this.onServerLeft.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.MESSAGE_CREATED, this.onMessageCreated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.MESSAGE_UPDATED, this.onMessageUpdated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.MESSAGE_DELETED, this.onMessageDeleted.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.MESSAGE_BUTTON_CLICKED, this.onMessageButtonClicked.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_ROLE_CREATED, this.onServerRoleCreated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_ROLE_DELETED, this.onServerRoleDeleted.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_ROLE_UPDATED, this.onServerRoleUpdated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_ROLE_ORDER_UPDATED, this.onServerRoleOrderUpdated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.MESSAGE_REACTION_ADDED, this.onMessageReactionAdded.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.MESSAGE_REACTION_REMOVED, this.onMessageReactionRemoved.bind(this));
    }
    onConnect() {
        this.socket.emit(EventNames_1.SocketClientEvents.AUTHENTICATE, {
            token: this.client.token,
        });
    }
    onAuthenticated(payload) {
        this.client.user = new ClientUser_1.ClientUser(this.client, payload.user);
        for (let i = 0; i < payload.servers.length; i++) {
            const server = payload.servers[i];
            this.client.servers.setCache(server);
        }
        for (let i = 0; i < payload.channels.length; i++) {
            const rawChannel = payload.channels[i];
            const channel = this.client.channels.setCache(rawChannel);
            if ("serverId" in rawChannel && rawChannel.serverId) {
                const server = this.client.servers.cache.get(rawChannel.serverId);
                server?.channels.set(channel.id, channel);
            }
        }
        for (let i = 0; i < payload.serverMembers.length; i++) {
            const member = payload.serverMembers[i];
            this.client.users.setCache(member.user);
            const server = this.client.servers.cache.get(member.serverId);
            server?.members.setCache(member);
        }
        for (let i = 0; i < payload.serverRoles.length; i++) {
            const role = payload.serverRoles[i];
            const server = this.client.servers.cache.get(role.serverId);
            server?.roles.setCache(role);
        }
        this.client.emit(EventNames_1.ClientEvents.Ready);
    }
    onServerMemberJoined(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        this.client.users.setCache(payload.member.user);
        const member = server?.members.setCache(payload.member);
        if (!member)
            return;
        this.client.emit(EventNames_1.ClientEvents.ServerMemberJoined, member);
    }
    onServerMemberUpdated(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const member = server?.members.cache.get(payload.userId);
        if (!member)
            return;
        updateClass(member, payload.updated);
        this.client.emit(EventNames_1.ClientEvents.ServerMemberUpdated, member);
    }
    onServerJoined(payload) {
        const server = this.client.servers.setCache(payload.server);
        for (let i = 0; i < payload.members.length; i++) {
            const member = payload.members[i];
            this.client.users.setCache(member.user);
            server?.members.setCache(member);
        }
        for (let i = 0; i < payload.roles.length; i++) {
            const role = payload.roles[i];
            server?.roles.setCache(role);
        }
        for (let i = 0; i < payload.channels.length; i++) {
            const channel = payload.channels[i];
            this.client.channels.setCache(channel);
        }
        this.client.emit(EventNames_1.ClientEvents.ServerJoined, server);
    }
    onServerChannelCreated(payload) {
        const channel = this.client.channels.setCache(payload.channel);
        const server = this.client.servers.cache.get(payload.serverId);
        server?.channels.set(channel.id, channel);
        this.client.emit(EventNames_1.ClientEvents.ServerChannelCreated, channel);
    }
    onServerChannelUpdated(payload) {
        const channel = this.client.channels.cache.get(payload.channelId);
        const updated = payload.updated;
        if (channel) {
            updateClass(channel, updated);
            this.client.emit(EventNames_1.ClientEvents.ServerChannelUpdated, channel);
        }
    }
    onServerChannelDeleted(payload) {
        const channel = this.client.channels.cache.get(payload.channelId);
        if (channel) {
            this.client.channels.cache.delete(payload.channelId);
            const server = this.client.servers.cache.get(payload.serverId);
            server?.channels.delete(payload.channelId);
            this.client.emit(EventNames_1.ClientEvents.ServerChannelDeleted, {
                channelId: payload.channelId,
                serverId: payload.serverId,
            });
        }
    }
    onServerLeft(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        if (!server)
            return;
        this.client.emit(EventNames_1.ClientEvents.ServerLeft, server);
        this.client.servers.cache.delete(payload.serverId);
        this.client.channels.cache.forEach((channel) => {
            if (channel instanceof ServerChannel_1.ServerChannel &&
                channel.serverId === payload.serverId) {
                this.client.channels.cache.delete(channel.id);
            }
        });
        server.members.cache.clear();
    }
    onServerMemberLeft(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const member = server?.members.cache.get(payload.userId);
        if (!member)
            return;
        this.client.emit(EventNames_1.ClientEvents.ServerMemberLeft, member);
        server?.members.cache.delete(payload.userId);
    }
    onMessageCreated(payload) {
        const message = this.client.messages.setCache(payload.message);
        this.client.emit(EventNames_1.ClientEvents.MessageCreate, message);
    }
    onMessageUpdated(payload) {
        const message = this.client.messages.cache.get(payload.messageId);
        const updated = payload.updated;
        if (message) {
            message._update(updated);
        }
        if (message) {
            this.client.emit(EventNames_1.ClientEvents.MessageUpdate, message);
        }
    }
    onMessageDeleted(payload) {
        this.client.messages.cache.delete(payload.messageId);
        this.client.emit(EventNames_1.ClientEvents.MessageDelete, {
            messageId: payload.messageId,
            channelId: payload.channelId,
        });
    }
    onMessageButtonClicked(payload) {
        const button = new MessageButton_1.MessageButton(this.client, payload);
        this.client.emit(EventNames_1.ClientEvents.MessageButtonClick, button);
    }
    onServerRoleCreated(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const role = server?.roles.setCache(payload);
        if (!role)
            return;
        this.client.emit(EventNames_1.ClientEvents.ServerRoleCreated, role);
    }
    onServerRoleDeleted(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const role = server?.roles.cache.get(payload.roleId);
        if (!role)
            return;
        server?.roles.cache.delete(payload.roleId);
        this.client.emit(EventNames_1.ClientEvents.ServerRoleDeleted, role);
    }
    onServerRoleUpdated(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const role = server?.roles.cache.get(payload.roleId);
        if (!role)
            return;
        updateClass(role, payload.updated);
        this.client.emit(EventNames_1.ClientEvents.ServerRoleUpdated, role);
    }
    onServerRoleOrderUpdated(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        for (let i = 0; i < payload.roleIds.length; i++) {
            const roleId = payload.roleIds[i];
            const role = server?.roles.cache.get(roleId);
            if (!role)
                continue;
            role.order = i + 1;
        }
        this.client.emit(EventNames_1.ClientEvents.ServerRoleOrderUpdated, server?.roles);
    }
    onMessageReactionAdded(payload) {
        const reaction = new Reaction_1.Reaction(this.client, payload);
        const user = this.client.users.cache.get(payload.reactedByUserId);
        this.client.emit(EventNames_1.ClientEvents.MessageReactionAdded, reaction, user);
    }
    onMessageReactionRemoved(payload) {
        const reaction = new Reaction_1.Reaction(this.client, payload);
        const user = this.client.users.cache.get(payload.reactionRemovedByUserId);
        this.client.emit(EventNames_1.ClientEvents.MessageReactionRemoved, reaction, user);
    }
}
function updateClass(classInstance, update) {
    for (const [key, value] of Object.entries(update)) {
        classInstance[key] = value;
    }
}
