// @nerimity/nerimity.js/build/classes/MessageButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageButton = void 0;
const MessageService_1 = require("../services/MessageService");
class MessageButton {
    constructor(client, payload) {
        this.client = client;
        this.id = payload.buttonId;
        this.userId = payload.userId;
        this.channelId = payload.channelId;
        this.messageId = payload.messageId;
        this.user = this.client.users.cache.get(this.userId);
        this.channel = client.channels.cache.get(this.channelId);
        this.data = payload.data;
        this.type = payload.type;
        this.partial = true;
        this.message = this.client.messages.cache.get(this.messageId);
        if (this.message) {
            this.partial = false;
        }
    }
    async fetch() {
        if (this.partial) {
            this.message = await this.client.messages.fetch(this.channelId, this.messageId);
            this.partial = false;
        }
    }
    async respond(opts) {
        await (0, MessageService_1.buttonClickCallback)({
            client: this.client,
            buttonId: this.id,
            channelId: this.channelId,
            messageId: this.messageId,
            userId: this.userId,
            data: opts,
        });
    }
}
exports.MessageButton = MessageButton;
