// @nerimity/nerimity.js/build/classes/Message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = exports.Messages = void 0;
const RawData_1 = require("../RawData");
const MessageService_1 = require("../services/MessageService");
const Collection_1 = require("./Collection");
const UserMentionRegex = /\[@:([0-9]+)\]/gm;
const CommandRegex = /^(\/[^:\s]*):(\d+)( .*)?$/m;
class Messages {
    constructor(client, limit) {
        this.client = client;
        this.cache = new Collection_1.Collection({ limit: limit ?? 1000 });
    }
    setCache(rawMessage) {
        const message = new Message(this.client, rawMessage);
        this.cache.set(message.id, message);
        return message;
    }
    async fetch(channelId, messageId, force = false) {
        if (!force) {
            const message = this.cache.get(messageId);
            if (message)
                return message;
        }
        const rawMessage = await (0, MessageService_1.fetchMessage)({
            channelId: channelId,
            client: this.client,
            messageId: messageId,
        });
        if (!rawMessage)
            return undefined;
        const message = new Message(this.client, rawMessage);
        this.cache.set(message.id, message);
        return message;
    }
}
exports.Messages = Messages;
class Message {
    constructor(client, message) {
        this.mentions = [];
        this.replies = new Collection_1.Collection();
        this.client = client;
        this.raw = message;
        this.id = message.id;
        this.channelId = message.channelId;
        this.channel = client.channels.cache.get(this.channelId);
        this.content = message.content;
        this.type = message.type;
        this.createdAt = message.createdAt;
        this.user = this.client.users.cache.get(message.createdBy.id);
        this.editedAt = message.editedAt;
        if (message.replyMessages?.length) {
            message.replyMessages.forEach((reply) => {
                if (reply.replyToMessage) {
                    this.replies.set(reply.replyToMessage.id, new Message(client, reply.replyToMessage));
                }
            });
        }
        const cmd = message.content?.match(CommandRegex);
        if (cmd?.[2] === this.client.user?.id) {
            this.command = {
                name: cmd?.[1].substring(1),
                args: message.content.split(" ").slice(1),
            };
        }
        if (!this.user) {
            this.user = this.client.users.setCache(message.createdBy);
        }
        if (!this.channel) {
            this.channel = this.client.channels.setCache({
                id: this.channelId,
                type: RawData_1.ChannelType.DM_TEXT,
            });
        }
        if (this.content) {
            this._updateMentions(this.content);
        }
    }
    get member() {
        return this.channel.server?.members.cache.get(this.user.id);
    }
    reply(content, _opts) {
        const opts = _opts || {};
        opts.replyToMessageIds = [this.id];
        return this.channel.send(content, opts);
    }
    async edit(content, opts) {
        const RawMessage = await (0, MessageService_1.editMessage)({
            client: this.client,
            channelId: this.channel.id,
            messageId: this.id,
            content: content,
            htmlEmbed: opts?.htmlEmbed,
            buttons: opts?.buttons,
        });
        const message = new Message(this.client, RawMessage);
        return message;
    }
    async delete() {
        return (0, MessageService_1.deleteMessage)({
            channelId: this.channel.id,
            client: this.client,
            messageId: this.id,
        });
    }
    _updateMentions(content) {
        const mentionIds = [...content.matchAll(UserMentionRegex)].map((exp) => exp[1]);
        this.mentions = mentionIds
            .map((id) => this.client.users.cache.get(id))
            .filter((u) => u);
    }
    _update(update) {
        this.content = update.content || this.content;
        this.editedAt = update.editedAt || this.editedAt;
        if (this.content) {
            this._updateMentions(this.content);
        }
        this.client.messages.cache.set(this.id, this);
    }
    toString() {
        return `[q:${this.id}]`;
    }
}
exports.Message = Message;
