// @nerimity/nerimity.js/build/classes/Post.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Post = exports.Posts = void 0;
const PostsService_1 = require("../services/PostsService");
class Posts {
    constructor(client) {
        this.client = client;
    }
    async get(id) {
        const RawPosts = await (0, PostsService_1.getPosts)(this.client);
        const posts = RawPosts.map((post) => new Post(this.client, post));
        return id ? posts.find((p) => p.id === id) : posts;
    }
    async create(content, opts) {
        const RawPost = await (0, PostsService_1.postPost)({
            client: this.client,
            content: content,
            nerimityCdnFileId: opts?.nerimityCdnFileId,
            poll: opts?.poll,
        });
        const post = new Post(this.client, RawPost);
        return post;
    }
}
exports.Posts = Posts;
class Post {
    constructor(client, post) {
        this.client = client;
        this.id = post.id;
        this.content = post.content;
        this.attachments = post.attachments;
        this.deleted = post.deleted;
        this.block = post.block;
        this.commentToId = post.commentToId;
        this.commentTo = post.commentTo;
        this.createdBy = post.createdBy;
        this.createdAt = post.createdAt;
        this.editedAt = post.editedAt;
        this.likedBy = post.likedBy;
        this.reposts = post.reposts;
        this.repost = post.repost;
        this._count = post._count;
        this.views = post.views;
        this.announcement = post.announcement;
        this.poll = post.poll;
    }
    async edit(content) {
        const RawPost = await (0, PostsService_1.editPost)({
            client: this.client,
            content: content,
            postId: this.id,
        });
        const post = new Post(this.client, RawPost);
        return post;
    }
    async delete() {
        await (0, PostsService_1.deletePost)({
            client: this.client,
            postId: this.id,
        });
    }
}
exports.Post = Post;
