// @nerimity/nerimity.js/build/classes/RPCClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RPCClient = void 0;
const eventemitter3_1 = __importDefault(require("eventemitter3"));
const ws_1 = require("ws");
const PORT_RANGES = [6463, 6472];
function findRunningServer(appId) {
    return new Promise((res) => {
        const servers = [];
        const id = setTimeout(() => {
            res(null);
        }, 5000);
        for (let port = PORT_RANGES[0]; port < PORT_RANGES[1] + 1; port++) {
            const server = new ws_1.WebSocket(`ws://localhost:${port}?appId=${appId}`);
            const index = servers.length;
            servers.push(server);
            server.on("message", (data) => {
                const message = jsonParseCatch(data.toString());
                if (message.name === "HELLO_NERIMITY_RPC") {
                    clearTimeout(id);
                    servers.forEach((server, i) => {
                        if (i === index)
                            return;
                        server.removeAllListeners();
                        server.close();
                    });
                    server.send(JSON.stringify(message));
                    res(server);
                }
            });
            server.on("error", () => { });
        }
    });
}
class RPCClient extends eventemitter3_1.default {
    constructor() {
        super();
        this.loggedIn = false;
    }
    send(payload) {
        if (!this.ws)
            return;
        this.ws.send(JSON.stringify({
            name: "UPDATE_RPC",
            data: payload,
        }));
    }
    async login(appId) {
        if (this.loggedIn)
            return true;
        this.appId = appId;
        this.ws = await findRunningServer(appId);
        if (!this.ws)
            return await this.login(appId);
        this.loggedIn = true;
        this.emit("ready");
        return true;
    }
}
exports.RPCClient = RPCClient;
const jsonParseCatch = (json) => {
    try {
        return JSON.parse(json);
    }
    catch {
        return null;
    }
};
