// @nerimity/nerimity.js/build/classes/ServerMember.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerMember = exports.ServerMembers = void 0;
const bitwise_1 = require("../bitwise");
const ServerService_1 = require("../services/ServerService");
const Collection_1 = require("./Collection");
class ServerMembers {
    constructor(client) {
        this.client = client;
        this.cache = new Collection_1.Collection();
    }
    setCache(rawMember) {
        const member = new ServerMember(this.client, rawMember);
        this.cache.set(member.user.id, member);
        return member;
    }
}
exports.ServerMembers = ServerMembers;
class ServerMember {
    constructor(client, member) {
        this.client = client;
        this.id = member.user.id;
        this.roleIds = member.roleIds;
        this.nickname = member.nickname;
        this.user = this.client.users.cache.get(member.user.id);
        this.server = this.client.servers.cache.get(member.serverId);
    }
    toString() {
        return `[@:${this.id}]`;
    }
    async ban(opts) {
        return (0, ServerService_1.banServerMember)(this.client, this.server.id, this.user.id, opts?.reason);
    }
    async unban() {
        return (0, ServerService_1.unbanServerMember)(this.client, this.server.id, this.user.id);
    }
    async kick() {
        return (0, ServerService_1.kickServerMember)(this.client, this.server.id, this.user.id);
    }
    get roles() {
        return this.roleIds
            .map((id) => this.server.roles.cache.get(id))
            .filter(Boolean);
    }
    permissions() {
        const defaultRoleId = this.server?.defaultRoleId;
        const defaultRole = this.server.roles.cache.get(defaultRoleId);
        let currentPermissions = defaultRole?.permissions || 0;
        const memberRoles = this.roles;
        for (let i = 0; i < memberRoles.length; i++) {
            const role = memberRoles[i];
            currentPermissions = (0, bitwise_1.addBit)(currentPermissions, role?.permissions || 0);
        }
        return currentPermissions;
    }
    hasPermission(permission, ignoreAdmin = false, ignoreCreator = false) {
        if (!ignoreCreator) {
            if (this.server.createdById === this.user.id)
                return true;
        }
        if (!ignoreAdmin) {
            if ((0, bitwise_1.hasBit)(this.permissions(), bitwise_1.RolePermissions.ADMIN))
                return true;
        }
        return (0, bitwise_1.hasBit)(this.permissions(), permission);
    }
}
exports.ServerMember = ServerMember;
