// @nerimity/nerimity.js/build/classes/Server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Server = exports.Servers = void 0;
const ServerService_1 = require("../services/ServerService");
const Collection_1 = require("./Collection");
const ServerMember_1 = require("./ServerMember");
const ServerRole_1 = require("./ServerRole");
class Servers {
    constructor(client) {
        this.client = client;
        this.cache = new Collection_1.Collection();
    }
    setCache(rawServer) {
        const server = new Server(this.client, rawServer);
        this.cache.set(server.id, server);
        return server;
    }
}
exports.Servers = Servers;
class Server {
    constructor(client, server) {
        this.channels = new Collection_1.Collection();
        this.client = client;
        this.id = server.id;
        this.name = server.name;
        this.avatar = server.avatar;
        this.members = new ServerMember_1.ServerMembers(this.client);
        this.roles = new ServerRole_1.ServerRoles(this.client);
        this.defaultRoleId = server.defaultRoleId;
        this.createdById = server.createdById;
    }
    async banMember(userId, opts) {
        return (0, ServerService_1.banServerMember)(this.client, this.id, userId, opts?.reason);
    }
    async unbanMember(userId) {
        return (0, ServerService_1.unbanServerMember)(this.client, this.id, userId);
    }
    async kickMember(userId) {
        return (0, ServerService_1.kickServerMember)(this.client, this.id, userId);
    }
}
exports.Server = Server;
